File: mobile/lib/data/repositories/in_memory_lesson_repository.dart
import '../../domain/repositories/i_lesson_repository.dart';
import '../../domain/entities/lesson.dart';
import '../../domain/entities/learning_path_node.dart';
import '../datasources/lessons_data.dart';
import '../datasources/roadmaps_data.dart';

class InMemoryLessonRepository implements ILessonRepository {
  @override
  Future<Lesson?> getLessonById(String lessonId) async {
    return seedLessons[lessonId];
  }

  @override
  Future<Lesson?> getLessonByCourseAndSkill(String courseId, String skillId) async {
    return seedLessons.values
        .where((l) => l.courseId == courseId && l.skillId == skillId)
        .firstOrNull;
  }

  @override
  Future<List<LearningPathNode>> getBaseRoadmapNodes(String courseId) async {
    return List.of(seedRoadmapNodes[courseId] ?? []);
  }
}

File: src/data/repositories/InMemoryLessonRepository.ts
import { ILessonRepository } from '@/domain/repositories/ILessonRepository';
import { Lesson } from '@/domain/entities/Lesson';
import { LearningPathNode } from '@/domain/entities/LearningPathNode';
import { SEED_LESSONS } from '../datasources/lessons';
import { SEED_ROADMAP_NODES } from '../datasources/roadmaps';

export class InMemoryLessonRepository implements ILessonRepository {
  async getLessonById(lessonId: string): Promise<Lesson | null> {
    const lesson = SEED_LESSONS[lessonId];
    return lesson ? { ...lesson } : null;
  }

  async getLessonByCourseAndSkill(courseId: string, skillId: string): Promise<Lesson | null> {
    const lesson = Object.values(SEED_LESSONS).find(
      (l) => l.courseId === courseId && l.skillId === skillId
    );
    return lesson ? { ...lesson } : null;
  }

  async getBaseRoadmapNodes(courseId: string): Promise<LearningPathNode[]> {
    const nodes = SEED_ROADMAP_NODES[courseId];
    return nodes ? nodes.map((n) => ({ ...n })) : [];
  }
}

export const inMemoryLessonRepository = new InMemoryLessonRepository();
